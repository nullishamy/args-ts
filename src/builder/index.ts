export * from './argument'
export * from './resolver'
export * from './command'
export * from './builtin'
export * from './default-resolvers'
export * as a from './default-arguments'
export * as builtin from './default-builtins'
