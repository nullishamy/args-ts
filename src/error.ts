export class ArgError extends Error {}

export class ParseError extends ArgError {
  static expected (expected: string, received: string, argString: string, index: number): ParseError {
    return new ParseError(`expected: ${expected}\nreceived: ${received}`, argString, index)
  }

  constructor (
    public readonly problem: string,
    public readonly argString: string,
    public readonly index: number
  ) {
    super(`${problem} @ ${index} : ${argString}`)
  }
}

export class CoercionError extends ArgError {
  constructor (
    public readonly expectedType: string,
    public readonly receivedValue: string,
    public readonly problem: string
  ) {
    super(`${problem}, expected '${expectedType}' received '${receivedValue}'`)
  }
}
export class CommandError extends ArgError {}
export class SchemaError extends ArgError {}

export class InternalError extends Error {}
