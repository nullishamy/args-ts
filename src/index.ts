import { usePlatform, nodePlatform } from './internal/platform'

export * from './args'
export * from './builder'
export * from './opts'
export * as util from './util'

usePlatform(nodePlatform)
