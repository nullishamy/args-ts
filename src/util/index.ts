export * from './helpers'
export * from './help'
export * from './completion'
export * from './argv'
export * from './logging'
