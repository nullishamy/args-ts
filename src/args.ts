import { Argument } from './builder'
import { ParseError } from './error'
import { tokenise, parse, matchValuesWithDeclarations, WrappedDeclaration } from './parser'
export interface Opts {
  name: string
  description: string
  unknownArgBehaviour: 'skip' | 'throw'
}

export class Args<TArgTypes = {
  [k: string]: boolean | string | number | undefined
}> {
  private declarations: Record<string, WrappedDeclaration> = {}

  constructor (private readonly opts: Opts) {}

  public add<
    // The declared argument type
    TArg,
    // The long declaration of the argument, used to setup the keys in the parsed object
    TLong extends string,
    // The return type of the custom argument function, used for inference
    TRet = never,
  >(
    // We don't need to care about the actual value of the short flag, we are only
    // going to provide the keys for the long variant. This is mostly because handling both keys,
    // particularly in the type system, is very annoying and overall very un-needed.
    [longFlag, shortFlag]: [`--${TLong}`, `-${string}`?],
    declaration: Argument<TArg>
  ): Args<TArgTypes & {
      // Syntax to add a key to a type
      [key in TLong]:
      // If we get passed a function, take its return type instead of the lookup table type
      // Then combined with the above syntax, add it to our arg types by the long key and the discovered value type
      TArg extends ((input: string) => never)
        ? TRet
        : TArg
    }> {
    if (!longFlag.startsWith('--')) {
      throw new ParseError(`long flags must start with '--', got '${longFlag}'`)
    }

    if (this.declarations[longFlag.substring(2)]) {
      throw new ParseError(`duplicate long flag '${longFlag}'`)
    }

    this.declarations[longFlag.substring(2)] = {
      inner: declaration,
      longFlag: longFlag.substring(2),
      shortFlag: shortFlag?.substring(1)
    }

    if (shortFlag) {
      if (!shortFlag.startsWith('-')) {
        throw new ParseError(`short flags must start with '-', got '${shortFlag}'`)
      }

      if (this.declarations[shortFlag.substring(1)]) {
        throw new ParseError(`duplicate short flag '${shortFlag}'`)
      }

      this.declarations[shortFlag.substring(1)] = {
        inner: declaration,
        longFlag: longFlag.substring(2),
        shortFlag: shortFlag.substring(1)
      }
    }

    // @ts-expect-error inferrence is broken
    return this
  }

  public async parse (argString: string): Promise<TArgTypes> {
    const tokens = tokenise(argString)
    const parsed = parse(tokens)
    const matched = await matchValuesWithDeclarations(parsed, this.declarations, this.opts)
    return Object.fromEntries([...matched.entries()].map(([key, value]) => [key.longFlag, value.parsed])) as TArgTypes
  }

  public reset (): void {
    this.declarations = {}
  }
}
