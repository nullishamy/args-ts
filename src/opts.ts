import { Resolver } from './builder'
import { EnvironmentResolver } from './builder/'
import { Logger } from './util'

type MakePassedOpts<TOpts, TDefaults extends keyof TOpts> = (
  & Omit<TOpts, TDefaults>
  & Partial<Pick<TOpts, TDefaults>>
)

export interface StoredParserOpts {
  /**
   * The program name, shown in the help text.
   */
  programName: string
  /**
   * The program description, shown in the help text.
   */
  programDescription: string
  /**
   * The program version, shown in the help text.
   */
  programVersion: string
  /**
   * What to do when an unrecognised argument (not defined in the schema) is passed in the user input.
   */
  unrecognisedArgument: 'skip' | 'throw'
  /**
   * What to do when an unrecognised command (not defined in the schema) is passed into the user input.
   */
  unrecognisedCommand: 'into-positional' | 'throw'
  /**
   * Whether to enable the "rest" syntax:
   * When enabled, the values are collected
   * ```
   * --flag value -x -- the rest value goes here
   * => {
   *  flag: "value",
   *  x: true,
   *  rest: "the rest value goes here"
   * }
   * ```
   * Otherwise, an error is thrown.
   * @remarks When this is enabled, it will overwrite any arguments with the long flag of "rest"
   */
  restSyntax: 'collect' | 'error'
  /**
   * What to do when a user runs a deprecated command.
   */
  deprecatedCommands: 'error' | 'unknown-command'
  /**
   * Whether to enable the "key = value" syntax:
   * When enabled, the values are parsed as if normal syntax was used
   * ```
   * --flag=value -x
   * => {
   *  flag: "value",
   *  x: true,
   * }
   * ```
   * Otherwise, an error is thrown.
   */
  keyEqualsValueSyntax: boolean
  /**
   * Whether to enable the "short flag group" syntax
   * When enabled, groups of short flags are parsed as if they had all been specified individually
   * ```
   * -xyz -ab
   * => {
   *  x: true,
   *  y: true,
   *  z: true,
   *  a: true,
   *  b: true
   * }
   * ```
   * Otherwise, a group is treated as a single argument.
   * @remarks This feature makes the assumption that any short flag with length \> 1 is a group, even if it maps to a real short flag
   */
  shortFlagGroups: boolean
  /**
   * The prefix to use when falling back to environment variables for missing arguments.
   */
  environmentPrefix: string | undefined
  /**
   * Whether users must provide at least one command.
   */
  mustProvideCommand: boolean
  /**
   * The prefix to use for "negated boolean" syntax. An empty string means this feature is disabled.
   * ```
   * --no-flag --other-flag
   * => {
   *  flag: false,
   *  ['other-flag']: true
   * }
   * ```
   */
  negatedBooleanPrefix: string
  /**
   * The logger used by the library to provide diagnostic information.
   */
  logger: Logger
  /**
   * The resolvers used by default. These are copied into the {@link ArgState} when a {@link Args} parser is constructed.
   */
  resolvers: Resolver[]
}

/**
 * The default parser options to use. Merged with the opts passed to the {@link Args} constructor.
 * Subject to change. These are opinionated defaults.
*/
// Default to the loudest possible error modes, to alert us of programming errors
export const defaultParserOpts = {
  unrecognisedArgument: 'throw',
  unrecognisedCommand: 'into-positional',
  deprecatedCommands: 'error',
  restSyntax: 'collect',
  shortFlagGroups: true,
  keyEqualsValueSyntax: true,
  environmentPrefix: undefined,
  mustProvideCommand: true,
  negatedBooleanPrefix: 'no-',
  logger: new Logger('default'),
  resolvers: [
    new EnvironmentResolver('env')
  ] as Resolver[]
} as const satisfies Partial<StoredParserOpts>

/**
 * @see {@link StoredParserOpts} for documentation
 */
export type ParserOpts = MakePassedOpts<StoredParserOpts, keyof typeof defaultParserOpts>

export interface StoredCommandOpts {
  /**
   * The description of the command. Used in the help text.
   */
  description: string
  /**
   * The parser options used for the command's parser.
   */
  parserOpts: StoredParserOpts

  /**
   * Whether this command is deprecated or not.
   */
  deprecated: boolean
  /**
   * The message to show if this command is deprecated, the command is run, and the {@link ParserOpts.deprecatedCommands} option is set to "error"
   */
  deprecationMessage: string
  /**
   * Whether this command is hidden from the help text.
   */
  hidden: boolean
}

/**
 * The default command options to use. Merged with the opts passed to the {@link Command} constructor.
 * Subject to change. These are opinionated defaults.
*/
export const defaultCommandOpts = {
  deprecated: false,
  deprecationMessage: 'This command is deprecated.',
  hidden: false
} as const satisfies Partial<StoredCommandOpts>

// Must override `parserOpts` in `StoredCommandOpts` so users can pass their single value around
/**
 * @see {@link StoredCommandOpts} for documentation
 */
export type CommandOpts = (
  & Omit<StoredCommandOpts, 'parserOpts' | keyof typeof defaultCommandOpts>
  & Partial<Pick<StoredCommandOpts, keyof typeof defaultCommandOpts>>
  & {
    parserOpts: ParserOpts
  }
)

export interface ArgumentOpts {
  /**
   * What to do when many arguments are passed to a type that only expects a single argument
   */
  tooManyArgs: 'drop' | 'throw'
  /**
   * What to do when an argument is specified multiple times. This is only for *non* array arguments.
   * @see arrayMultipleDefinitions for array argument behaviour
   */
  tooManyDefinitions: 'drop' | 'throw' | 'overwrite'
  /**
   * What to do when an argument is specified multiple times. This is only for array arguments.
   * @see arrayMultipleDefinitions for single argument behaviour
   */
  arrayMultipleDefinitions: 'append' | 'drop' | 'throw' | 'overwrite'
}

/**
 * The default argument options to use. Set as the default when an {@link Argument} is constructed.
 * Subject to change. These are opinionated defaults.
*/
export const defaultArgumentOpts = {
  tooManyArgs: 'throw',
  tooManyDefinitions: 'throw',
  arrayMultipleDefinitions: 'append'
} as const satisfies Partial<ArgumentOpts>
