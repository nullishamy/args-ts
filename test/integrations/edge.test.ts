import { Args, a } from '../../src'
import { parserOpts } from '../shared'
import { readFile } from 'fs/promises'
import { CoercionError, ParseError } from '../../src/error'
import path from 'path'

describe('Edge cases', () => {
  const parser = new Args(parserOpts)
    .arg(['--non-latin'], a.string())

  // https://github.com/minimaxir/big-list-of-naughty-strings/blob/master/blns.json
  it('can parse the naughty strings', async () => {
    const strings: string[] = JSON.parse(
      await readFile(
        path.join(__dirname, 'naughty-strings.json'),
        { encoding: 'utf-8' }
      )
    )

    const warnings: Error[] = []
    for (const naughty of strings) {
      let result: Awaited<ReturnType<typeof parser['parseToResult']>>

      try {
        // Try various string formations to induce errors
        await parser.parseToResult(`--non-latin ${naughty}`)
        await parser.parseToResult(`--non-latin '${naughty}'`)
        result = await parser.parseToResult(`--non-latin "${naughty}"`)
      } catch (err) {
        throw new Error(`error when parsing '${naughty}':\n${(err as Error).stack}`)
      }

      // "errors" are acceptable here, so long as the parser does not throw
      // log them so that we can spot strange messages
      if (!result.ok) {
        if (Array.isArray(result.err)) {
          result.err.forEach(e => warnings.push(e))
          continue
        }

        warnings.push(result.err)
        continue
      }

      if (result.val.mode !== 'args') {
        throw new Error(`result was not args, got ${result.val.mode} for '${naughty}'`)
      }

      expect(result.val.args['non-latin']).toStrictEqual(naughty)
    }

    let warnLog = ''
    for (const warning of warnings) {
      if (warning instanceof CoercionError) {
        warnLog += `coercion: ${warning.problem}`
      } else if (warning instanceof ParseError) {
        warnLog += `parse: ${warning.problem}`
      } else {
        warnLog += `unknown error: ${warning.message}`
      }

      warnLog += '\n'
    }

    // We will want to inspect any changes to our logs, in case we accidentally broke something
    // but we won't want to do it ourselves, have jest do the diffing and updating for us.
    expect(warnLog).toMatchSnapshot()
  })
})
