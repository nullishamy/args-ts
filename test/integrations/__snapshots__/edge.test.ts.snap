// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge cases can parse the naughty strings 1`] = `
"coercion: argument 'non-latin' is not declared as optional, does not have a default, and was not provided a value
coercion: excess argument(s) to --non-latin: '{}|_+"'
coercion: argument 'non-latin' is not declared as optional, does not have a default, and was not provided a value
parse: expected: flag -- received: {"type":"quote","value":"\\""}
coercion: excess argument(s) to --non-latin: '"'
parse: expected: flag -- received: {"type":"quote","value":"'"}
parse: expected: flag -- received: {"type":"quote","value":"'"}
parse: expected: flag -- received: {"type":"char","value":">"}
parse: expected: flag -- received: {"type":"char","value":"o"}
parse: expected: flag -- received: {"type":"char","value":";"}
parse: expected: flag -- received: {"type":"char","value":">"}
parse: expected: flag -- received: {"type":"char","value":"a"}
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: 'text/javascript">javascript:alert(1);</script>"'
coercion: excess argument(s) to --non-latin: '><\\x3Cscript>javascript:alert(1)</script>"'
coercion: excess argument(s) to --non-latin: '><\\x00script>javascript:alert(1)</script>"'
coercion: excess argument(s) to --non-latin: 'x\\x3Aexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:expression\\x5C(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:expression\\x00(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:exp\\x00ression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:exp\\x5Cression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x0Aexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x09expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE3\\x80\\x80expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x84expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xC2\\xA0expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x80expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x8Aexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x0Dexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x0Cexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x87expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xEF\\xBB\\xBFexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x20expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x88expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x00expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x8Bexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x86expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x85expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x82expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\x0Bexpression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x81expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x83expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: 'x:\\xE2\\x80\\x89expression(javascript:alert(1)">DEF"'
coercion: excess argument(s) to --non-latin: '\\x0Bjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x0Fjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xC2\\xA0javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x05javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE1\\xA0\\x8Ejavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x18javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x11javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x88javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x89javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x80javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x17javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x03javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x0Ejavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Ajavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x00javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x10javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x82javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x20javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x13javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x09javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x8Ajavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x14javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x19javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\xAFjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Fjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x81javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Djavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x87javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x07javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE1\\x9A\\x80javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x83javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x04javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x01javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x08javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x84javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x86javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE3\\x80\\x80javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x12javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x0Djavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x0Ajavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x0Cjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x15javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\xA8javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x16javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x02javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Bjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x06javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\xA9javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x80\\x85javascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Ejavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\xE2\\x81\\x9Fjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '\\x1Cjavascript:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: 'javascript\\x00:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: 'javascript\\x3A:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: 'javascript\\x09:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: 'javascript\\x0D:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: 'javascript\\x0A:javascript:alert(1)"', 'id="fuzzelement1">test</a>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x0Aonerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x22onerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x0Bonerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x0Donerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x2Fonerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x09onerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x0Conerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x00onerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x27onerror=javascript:alert(1)>"'
coercion: excess argument(s) to --non-latin: '><img src=xxx:x \\x20onerror=javascript:alert(1)>"'
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
parse: expected: flag -- received: {"type":"char","value":"\`"}
coercion: excess argument(s) to --non-latin: 'alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'javascript:alert(1)">"'
coercion: excess argument(s) to --non-latin: 'x\`', '\`<script>javascript:alert(1)</script>"\`', '\`>"'
coercion: excess argument(s) to --non-latin: '"= alt=javascript:alert(1)//">"'
coercion: excess argument(s) to --non-latin: '\`><img', 'src=x:x', 'onerror=javascript:alert(1)></a>">"'
coercion: excess argument(s) to --non-latin: '/\\%(jscript)s"></script>"'
coercion: excess argument(s) to --non-latin: '\\\\%(jscript)s"></script>"'
parse: expected: flag -- received: {"type":"char","value":">"}
coercion: excess argument(s) to --non-latin: 'alert('xxs')">"'
coercion: excess argument(s) to --non-latin: 'alert('xxs')">"'
coercion: excess argument(s) to --non-latin: 'alert('xxs')">"'
coercion: excess argument(s) to --non-latin: 'jav', 'ascript:alert('XSS');">"'
coercion: excess argument(s) to --non-latin: 'jav&#x09;ascript:alert('XSS');">"'
coercion: excess argument(s) to --non-latin: 'jav&#x0A;ascript:alert('XSS');">"'
coercion: excess argument(s) to --non-latin: 'jav&#x0D;ascript:alert('XSS');">"'
coercion: excess argument(s) to --non-latin: '<IMG', 'SRC=java\\0script:alert(\\"XSS\\")>";'', '>', 'out"'
coercion: excess argument(s) to --non-latin: '&#14;', 'javascript:alert('XSS');">"'
coercion: excess argument(s) to --non-latin: 'http://ha.ckers.org/xss.js"></SCRIPT>"'
coercion: excess argument(s) to --non-latin: 'XSS")>"'
coercion: excess argument(s) to --non-latin: 'http://ha.ckers.org/xss.js"></SCRIPT>"'
coercion: excess argument(s) to --non-latin: 'XSS");//<</SCRIPT>"'
coercion: excess argument(s) to --non-latin: 'javascript:alert('XSS')""'
coercion: excess argument(s) to --non-latin: ';alert('XSS');//"'
coercion: excess argument(s) to --non-latin: 'touch', '/tmp/blns.fail"]}"'
coercion: excess argument(s) to --non-latin: 'puts', 'hello world', ')'
coercion: excess argument(s) to --non-latin: 'ls'
coercion: excess argument(s) to --non-latin: 'ls'
coercion: excess argument(s) to --non-latin: '1.0"', 'encoding="ISO'
coercion: excess argument(s) to --non-latin: '.__class__.__mro__[2].__subclasses__()[40](', '/etc/passwd").read()', '}}"'
"
`;
